### DISCUSSION TITLE
Tutorial request: k3s for developers, from container to scale (on own servers, without k8s knowledge)

### DISCUSSION THREAD

**Discussion by bluepuma77** [k3s-io/k3s#3396] [Ideas]
Dear k3s team,

it would be great to get a developer friendly tutorial how to use k3s and its preferred tools to move from a dev container to a production cluster. I am looking for a bare metal ("dedicated server" with linux) simple command line tutorial without the need of learning Terraform, Vagrant, Ansible, Helm, BGP and so on and so forth.
 
Currently we have an external high-available load-balancer, 3 ingress servers running Traefik and Docker Swarm Masters, and multiple worker nodes. We spin up 3+ containers for each tennant, they automatically register with Traefik and are available load-balanced via their sub-domain.

![k3s-example](https://user-images.githubusercontent.com/5221187/120620292-21065d80-c45d-11eb-860c-49e9d36e7e37.png)

Docker Swarm has served us well as a very easy to use tool to get started, but we seem to have some minor issues with high loads. At the same time no-one seems to be using Docker Swarm nowadays, related GitHub projects are mostly dormant. K3s seems to be a good alternative, especially as the initial setup seems as easy as Docker Swarm.

What I haven't found so far is an overarching tutorial of how to scale a web container with k3s on own servers. Yes, it is k8s compatible, but at the same time it has some tools build in and/or preferred, so a dedicated tutorial would be helpful. And it would be great to NOT have a lot of external tools which need to be understood and maintained. For starters it's just easier to setup a new server every month by hand instead of learning yet another tool to do this automatically. And for GDPR and money saving reasons our infrastructure is on dedicated servers, even [A18z recognised cloud is expensive](https://a16z.com/2021/05/27/cost-of-cloud-paradox-market-cap-cloud-lifecycle-scale-growth-repatriation-optimization/). To make live easier our scenario just uses an external load balancer.

It should include (using Docker Swarm terms here)
- [x] setup of k3s cluster
- [ ] setup of ingress with traefik and (purchased) SSL certificates (on 3 servers with fixed IP:port for high availability)
- [ ] start a service with x containers with sub-domain as label and node-constraint (on workers only)
- [ ] view status of cluster (maybe with a GUI)
- [ ] do a rolling update of a service
- [ ] advanced: collect logs with Prometheus/Grafana or Beats/Elastic/Kibana
- [ ] more advanced: make it high-available without external LB but with external Floating IP

For reference a post which gets Traefik and a web service up and running on Docker Swarm [with two config files and 3 CLI commands](https://www.reddit.com/r/docker/comments/nf7dfo/how_to_setup_docker_swarm_traefik_24_domainbased/gyyhws1?utm_source=share&utm_medium=web2x&context=3).

**Comment by dereknola** [comment]
Have you looked at the Rancher Academy course on k3s? 
https://community.suse.com/courses/4522316?utm_source=manual

**Reply by bluepuma77** [reply]
Thanks, I haven't done a course yet, but I briefly looked at it. 

It might be a solution, but 
a) I don't want to watch hours of videos when I just need a simple sample yaml file to deploy my app
b) I don't want to watch hours of videos when it explains things like how to use a different router when traefik is already there

**Comment by bluepuma77** [comment]
Every now and then I still think about kubernetes and using k3s. It sounds like a great proposition, having all-in-one, easy setup, reverse proxy ready. If the was only a simple BEST PRACTICE K3S document. Why is there no single article describing how to deploy my web-service with k3s/traefik and scale it to multiple nodes?

When searching for k3s and traefik you find 10s of articles - that explain how they DISABLE traefik and use another proxy instead. Why is traefik bundled with k3s when seemingly nobody is using it? Why is there not a single article how to make use of the k3s preferred choice of tools? Am I missing something?
